lib/Consul/Check.pm

```diff
@@ -3,20 +3,56 @@ package Consul::Check;
 use namespace::autoclean;
 
 use Moo::Role;
-use Types::Standard qw(Str HashRef);
+use Types::Standard qw(Str HashRef Enum);
 
-has name  => ( is => 'ro', isa => Str, required => 1 );
-has id    => ( is => 'ro', isa => Str );
-has notes => ( is => 'ro', isa => Str );
+use Carp qw(croak);
+
+my $status_type = Enum['passing', 'warning', 'critical'];
+
+has name       => ( is => 'ro', isa => Str, required => 1 );
+has id         => ( is => 'ro', isa => Str );
+has notes      => ( is => 'ro', isa => Str );
+has service_id => ( is => 'ro', isa => Str );
+has status     => ( is => 'ro', isa => $status_type );
+has deregister_critical_service_after =>
+                  ( is => 'ro', isa => Str );
+
+sub smart_new {
```

Comment:
Honestly, this smells a bit. It seems like the calling code already have some idea of the check type, because its adding check-specific args. Or do you have an untyped store of checks or something, and you just want to drop whatever you happen to have straight in? Either way, it feels to me like the caller should be explicit about what it wants. But I could be wrong.

I'm not keen on calling `Moo::Object->BUILDARGS` directly, because its undocumented/internal. I see why, you want it flatten the input hash in whatever form it happens to be in. It just feels wrong.

Incidentally, `Type::Params::multisig` is what I use these days to declare "args can look like A, B or C".